### PR TITLE
Begin of the translate mode

### PR DESCRIPTION
**Pull request by loco-odoo**

### CODE REVIEW

File: addons/html_builder/static/src/website_builder_action.js [hunk 1]
```diff
@@ -97,10 +125,90 @@ class WebsiteBuilder extends Component {
             targetDoc: this.websiteContent.el.contentDocument,
         });
         this.websiteService.pageDocument = this.websiteContent.el.contentDocument;
+        if (this.translation) {
```
Comment (loco-odoo) [inline addons/html_builder/static/src/website_builder_action.js:128]: Is it possible that `this.translation` is not set when arriving here: in other world, is it possible that `onIframeLoad` is called before `onPatched` ?

File: addons/html_builder/static/src/builder/snippets_menu.js [hunk 2]
```diff
@@ -230,6 +221,25 @@ export class SnippetsMenu extends Component {
...
+            translations[this.websiteService.currentWebsite.metadata.lang] = {
+                [el.dataset["oeTranslationSourceSha"]]: el.innerHTML,
+            };
+            return this.orm.call(el.dataset["oeModel"], "web_update_field_translations", [
+                [Number(el.dataset["oeId"])],
+                el.dataset["oeField"],
+                translations,
+            ]);
+        }
+        return this.saveView(el);
```
Comment (loco-odoo) [inline addons/html_builder/static/src/builder/snippets_menu.js:250]: For the moment, this is only useful for image fields. One big limitation for the moment is that in translate mode, although the option previews work correctly, it is currently not possible to save something else than text change (except an image field change).

File: addons/html_builder/static/tests/builder_action.test.js [hunk 3]
```diff
@@ -10,3 +11,20 @@ test("trigger mobile view", async () => {
...
+            this.props.action.context = {
+                params: {
+                    path: "/web/content/",
+                },
+            };
+            super.setup();
+        },
+    });
+    await setupWebsiteBuilder(`<h1> Homepage </h1>`);
+    expect(websiteBuilder.initialUrl).toBe("/website/force/1?path=%2F");
```
Comment (loco-odoo) [inline addons/html_builder/static/tests/builder_action.test.js:29]: I made a test here but I do not really like it: it does not bring much value. To be discussed

File: addons/html_builder/static/src/builder/snippets_menu.js [hunk 4]
```diff
@@ -96,6 +88,8 @@ export class SnippetsMenu extends Component {
                         this.state.currentOptionsContainers = currentOptionsContainers;
                         this.setTab("customize");
                     },
+                    unsplittable_node_predicates: (node) =>
+                        node.querySelector("[data-oe-translation-source-sha]"),
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/builder/snippets_menu.js:95]: Maybe just add it in translation mode ? (Or add a comment) We maybe want use another config in translation ?

Comment (FrancoisGe) [reply addons/html_builder/static/src/builder/snippets_menu.js:95]: remove few plugins etc ?

File: addons/html_builder/static/src/builder/snippets_menu.js [hunk 5]
```diff
@@ -38,24 +38,16 @@ const BUILDER_PLUGIN = [
...
 // todo: Why is it called SnippetsMenu? Should we rename it to BuilderSidebar?
 export class SnippetsMenu extends Component {
     static template = "html_builder.SnippetsMenu";
     static components = { BlockTab, CustomizeTab, InvisibleElementsPanel };
     static props = {
-        iframe: { type: Object },
         closeEditor: { type: Function },
         snippetsName: { type: String },
         toggleMobile: { type: Function },
+        translation: { type: Boolean },
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/builder/snippets_menu.js]: isTranslation ?

File: addons/html_builder/static/src/builder/snippets_menu.js (hunk 2 above)
Comment (FrancoisGe) [reply addons/html_builder/static/src/builder/snippets_menu.js:250]: It s the same thing in master ?

File: addons/html_builder/static/src/builder/snippets_menu.xml [hunk 6]
```diff
@@ -15,9 +15,15 @@
             </div>
         </div>
         <div class="o-snippets-tabs d-flex justify-content-between mt-2 p-2 text-uppercase flex-shrink-0">
-            <span class="px-2 cursor-pointer pb-1 pe-auto" t-att-class="{'active': state.activeTab === 'blocks'}" t-on-click="() => this.setTab('blocks')">Blocks</span>
-            <span class="px-2 cursor-pointer pb-1" t-att-class="{'active': state.activeTab === 'customize'}" t-on-click="() => this.setTab('customize')">Customize</span>
-            <span class="px-2 cursor-pointer pb-1" t-att-class="{'active': state.activeTab === 'theme'}" t-on-click="() => this.setTab('theme')">Theme</span>
+            <button class="px-2 cursor-pointer pb-1 pe-auto bg-transparent text-white border-0" t-att-class="{'active': state.activeTab === 'blocks'}" t-on-click="() => this.setTab('blocks')" t-att-disabled="props.translation">
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/builder/snippets_menu.xml]: why replace span by button ?

File: addons/html_builder/static/src/website_builder_action.js [hunk 7]
```diff
@@ -29,9 +37,13 @@ class WebsiteBuilder extends Component {
         });
         this.state = useState({ isEditing: false, isMobile: false });
         this.websiteService = useService("website");
+        this.ui = useService("ui");
         // TODO: to remove: this is only needed to not use the website systray
         // when using the "website preview" app.
         this.websiteService.useMysterious = true;
+        this.translation =
+            !!this.props.action.context.params &&
+            !!this.props.action.context.params.edit_translations;
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_builder_action.js]:
```suggestion
        this.translation =
            !!this.props.action.context.params?.edit_translations
```

File: addons/html_builder/static/src/website_builder_action.js [hunk 8]
```diff
@@ -40,9 +52,22 @@ class WebsiteBuilder extends Component {
                 this.websiteService.fetchUserGroups(),
             ]);
             this.backendWebsiteId = unslugHtmlDataObject(backendWebsiteRepr).id;
-            this.initialUrl = `/website/force/${encodeURIComponent(this.backendWebsiteId)}`;
+            const encodedPath = encodeURIComponent(this.path);
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_builder_action.js:60]: ?

File: addons/html_builder/static/src/website_builder_action.js [hunk 9]
```diff
@@ -40,9 +52,22 @@ class WebsiteBuilder extends Component {
...
+            this.initialUrl = `/website/force/${encodeURIComponent(
+                this.backendWebsiteId
+            )}?path=${encodedPath}`;
             this.websiteService.currentWebsiteId = this.backendWebsiteId;
         });
+        onMounted(() => {
+            const edition =
+                this.props.action.context.params &&
+                (!!this.props.action.context.params.enable_editor ||
+                    !!this.props.action.context.params.edit_translations);
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_builder_action.js]:
```suggestion
            const {enable_editor,  edit_translations} = this.props.action.context.params || {};
            const edition = !!(enable_editor || edit_translations);
```

File: addons/html_builder/static/src/website_builder_action.js [hunk 10]
```diff
@@ -97,10 +121,86 @@ class WebsiteBuilder extends Component {
             targetDoc: this.websiteContent.el.contentDocument,
         });
         this.websiteService.pageDocument = this.websiteContent.el.contentDocument;
+        if (this.translation) {
+            deleteQueryParam("edit_translations", this.websiteService.contentWindow, true);
+        }
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_builder_action.js:130]: ? It s for the reload ?

File: addons/html_builder/static/src/website_builder_action.js [hunk 11]
```diff
@@ -97,10 +121,86 @@ class WebsiteBuilder extends Component {
...
+
     toggleMobile() {
         this.state.isMobile = !this.state.isMobile;
     }
 }
 
+function deleteQueryParam(param, target = window, adaptBrowserUrl = false) {
+    const url = new URL(target.location.href);
+    url.searchParams.delete(param);
+    target.history.replaceState(target.history.state, null, url);
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_builder_action.js:185]: we should maybe do it in the action service no ?

File: addons/html_builder/static/src/website_builder_action.xml [hunk 12]
```diff
@@ -12,7 +12,7 @@
             </div>
         </div>
         <div  t-attf-style="width:{{state.isEditing ? 288 : 0}}px" class="o-website-snippetsmenu border-start border-dark">
-            <LazyComponent  t-if="state.isEditing" Component="'website.SnippetsMenu'" props="menuProps" bundle="'website.assets_builder'"/>
+            <LazyComponent  t-if="state.isEditing" Component="'website.SnippetsMenu'" props="this.menuProps" bundle="'website.assets_builder'"/>
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_builder_action.xml:16]: did you find the why you need to add the this ?

File: addons/html_builder/static/src/website_systray_item.js [hunk 13]
```diff
@@ -1,9 +1,85 @@
...
+            this.websiteService.currentWebsite &&
+            this.websiteService.currentWebsite.metadata.translatable
+        );
+    }
+
+    async attemptStartTranslate() {
+        // TODO: move on the website part (not html_builder) and add a test tour
+        if (this.websiteService.isRestrictedEditor && !this.websiteService.isDesigner) {
+            const object = this.websiteService.currentWebsite.metadata.mainObject;
+            const objects = {
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_systray_item.js]: objects is a strange name ^^ Maybe find something more explicit

File: addons/html_builder/static/src/website_systray_item.xml [hunk 14]
```diff
@@ -4,7 +4,34 @@
 <t t-name="html_builder.WebsiteSystrayItem">
     <div class="d-flex">
         <a href="#" class="o-website-btn-custo-secondary btn d-flex align-items-center rounded-0 border-0 px-3" t-on-click="props.onNewPage">New</a>
-        <a href="#" class="o-website-btn-custo-primary btn d-flex align-items-center rounded-0 border-0 px-3" t-on-click="props.onEditPage">Edit</a>
+        <a t-if="!translatable and !this.websiteService.is404" href="#" class="o-website-btn-custo-primary btn d-flex align-items-center rounded-0 border-0 px-3" t-on-click="props.onEditPage">Edit</a>
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_systray_item.xml]: Could you show me the edit 404 business flow ?

File: addons/html_builder/static/src/website_systray_item.xml [hunk 15]
```diff
@@ -4,7 +4,34 @@
...
+        <Dropdown t-else="">
+            <button class="o-website-btn-custo-primary btn rounded-0 border-0 px-3" data-hotkey="a">Edit</button>
+            <t t-set-slot="content">
+                <t t-if="this.websiteService.is404">
```
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_systray_item.xml]: We don t observe the is404 ? It s not reactive ?

File: addons/html_builder/static/src/website_systray_item.xml (hunk 14 above)
Comment (FrancoisGe) [inline addons/html_builder/static/src/website_systray_item.xml]: Maybe move all the translation logic in 1 component ? EditWebsiteSystrayItem

File: addons/html_builder/static/tests/builder_action.test.js (hunk 3 above)
Comment (FrancoisGe) [reply addons/html_builder/static/tests/builder_action.test.js:29]: We can discuss in vocal

File: addons/html_builder/static/tests/translation.test.js [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+        },
+    },
+};
+
+test("systray in translate mode", async () => {
+    patchWithCleanup(WebsiteSystrayItem.prototype, {
+        setup() {
+            this.websiteService = websiteServiceInTranslateMode;
+        },
+    });
```
Comment (FrancoisGe) [inline addons/html_builder/static/tests/translation.test.js]: Why not replace it in the service registry ?

File: addons/html_builder/static/tests/translation.test.js [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+            </p>
+        </div>`;
+}
+
+function setupSnippetsMenuForTranslation() {
+    patchWithCleanup(SnippetsMenu.prototype, {
+        setup() {
+            this.props.translation = true;
+            super.setup();
+            this.websiteService = websiteServiceInTranslateMode;
```
Comment (FrancoisGe) [inline addons/html_builder/static/tests/translation.test.js:120]: replace in service registry no ?

File: addons/html_builder/static/tests/translation.test.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+            <p>
+                <span data-oe-model="ir.ui.view" data-oe-id="526" data-oe-field="arch_db" data-oe-translation-state="to_translate" data-oe-translation-source-sha="sourceSha" class="o_editable">${inWrap}</span>
+            </p>
+        </div>`;
+}
+
+function setupSnippetsMenuForTranslation() {
+    patchWithCleanup(SnippetsMenu.prototype, {
+        setup() {
+            this.props.translation = true;
```
Comment (FrancoisGe) [inline addons/html_builder/static/tests/translation.test.js]: I m not a big fan to edit props
It s not possible to patch the parent to be true ?
(If something trigger an update props, you will have a strange issue)

**Pull request merged**